catkin_ws/src/state_estimation/src/ekf.py

```diff
@@ -0,0 +1,391 @@
+    rpy_vec = quaternion_to_euler(q_vec)
+    return rpy_vec_to_msg(rpy_vec)
+
+def quaternion_to_euler(quat_vec):
+        # From wikipedia (https://en.wikipedia.org/wiki/Conversion_between_quaternions_and_Euler_angles)
+
+        w = quat_vec[0]
+        x = quat_vec[1]
+        y = quat_vec[2]
+        z = quat_vec[3]
+
+        angles = np.array([0, 0 ,0])
+
+        # roll (x-axis rotation)
+        sinr_cosp = +2.0 * (w * x + y * z)
+        cosr_cosp = +1.0 - 2.0 * (x * x + y * y)
+        #angles[0] = np.arctan2(sinr_cosp, cosr_cosp)
+        
+
+        # yaw (z-axis rotation)
+        siny_cosp = +2.0 * (w * z + x * y)
+        cosy_cosp = +1.0 - 2.0 * (y * y + z * z)
+        #angles[2] = math.atan2(siny_cosp, cosy_cosp)
+
+        # pitch (y-axis rotation)
+        sinp = +2.0 * (w * y - z * x)
+        if np.abs(sinp) >= 1:
+            #angles[1] = np.sign(sinp) * math.pi # use 90 degrees if out of range
+            return np.array([ np.arctan2(sinr_cosp, cosr_cosp),
+                              np.sign(sinp) * math.pi,
+                              np.arctan2(siny_cosp, cosy_cosp)
+                              ])
+        else:
+            #angles[1] = math.asin(sinp)
+            return np.array([ np.arctan2(sinr_cosp, cosr_cosp),
+                              np.arcsin(sinp),
+                              np.arctan2(siny_cosp, cosy_cosp)
+                              ])
+
+        return angles
+
+
+quat_to_euler_jacobian = jacobian(quaternion_to_euler)
+
+
+
+class EKF:
+    def __init__(self, config):
+        # https://en.wikipedia.org/wiki/Extended_Kalman_filter
+
+        ''' Model Description
+        x = state
+        u = inputs (thruster forces, etc)
+        z = ouputs (measurements)
+        P = unvertainty/variance matrix of state x
+
+        Dynamics: x[k+1] = f(x[k], u[k])
+        Outputs:    z[k] = h(x[k], u[k])
+
+        Linear Form:
+          x[k+1] = A*x[k] + B*u[k]
+            z[k] = C*x[k] + B*u[k]
+        '''
+
+        self.n = 13 # Number of state elements
+        self.m = 8 # Number of inputs
+
+        self.x = np.zeros((self.n, 1))
+        self.x[IDx.Ow] = 1
+
+        self.u = np.zeros((self.m,1))
+
+        self.B = np.array(config.get_thrusts_to_wrench_matrix())
+
+        self.P = np.eye(self.n)
+        self.Q = np.eye(self.n) * 0.01 # Uncertanty in dynamics model
+
+        self.depth_sub = PressureSensorListener()
+        self.imu_sub = IMUSensorListener()
+
+        self.rate = 20
+        self.rate_ctrl = rospy.Rate(self.rate)
+
+        self.calc_F = jacobian(self.f)
+
+        self.sub_state_msg = SubState()
+        self.state_pub = rospy.Publisher("state_estimation", SubState, queue_size=1)
+
+        self.motor_sub = rospy.Subscriber("motor_command", MotorControls, self.motor_cb)
+
+    def motor_cb(self, msg):
+        self.u = np.array(msg.motorThrusts)
+
+    def prediction(self):
+
+        # Get jacobian of function f
+        F = self.calc_F(self.x, self.u)
```

Comment:
Potential issue: only the most recent motor command is taken into account. If for whatever reason the motor commands change during the course of the time step, then the previous motor command's values will be ignored.
Not sure if this will actually have a measurable effect or not, or if this is actually a scenario that may happen.

Reply:
I think there will be some cases where that happens, but it likely won't be a huge issue on these timescales.

A nice thing about the EKF is that we can sort of account for potential issues like this via the uncertainty we assign to the dynamics function.